### ISSUE TITLE
Glossary for naming APIs

### ISSUE THREAD

**Issue by encukou** [capi-workgroup/api-evolution#23]
*Parts taken fromVictor's suffix survey: https://github.com/capi-workgroup/problems/issues/52 *

We might want to provide a glossary to help keep API names consistent. Here's a draft.

## Name components

- `From<type>` (e.g. `PyLong_FromLong`): creates an object
- `As<type>` (e.g. `PyLong_AsLong`): converts to a C type
- `Get`: Retrieve a value (no side effect except allocation/refcounting)
- `Va`: Takes varargs

## Suffixes

- `Object` (in functions, e.g. ): Takes a Python object rather than a native type (e.g. `PyImport_AddModuleObject(PyObject*)` vs. `PyImport_AddModule(char*)`)
- `Object` (type, e.g. `PyLongObject`): The instance struct. Omit the underscore (it's not `PyLong_Object`).
- `_<typename>` (`_PyObject_CallFunction_SizeT`): Has an unexpected argument/return type, or a different type than a similar function
- `_impl`: "Inner" function that has a wrapper. Used e.g. by Argument Clinic, and by C-specific speedups (#18). Should be marked private (`_Py`) if exposed to users.

### Exceptions to rules

(These are new proposals)

- `Borrow`: Returns a borrowed reference
- Takes ownership of an argument:
  - `Steal`: used for `PyObject*` refcounted/GCd (other references might exist)
  - `Consume`: used for non-refcounted data (e.g. array of `PyObject*`)
  - `Clear`: Destroy a reference and mark it invalid (i.e. NULL it)
  - `Decref`, `Close`, `Free`, `Dealloc`: used for specific API that *only* destroys the reference/data
- `Static`: Takes a static immutable argument (e.g. a docstring that the VM doesn't need to copy)
- `Unchecked`: Doesn't validate a (pre)condition that's normally checked
- `Unsafe`: Doesn't uphold some invariant that users expect

### Suffixes for "evolved" versions of functions:

- `Ref`: Returns a new reference
- `WithError`: Has proper error handling
- `With*` (`WithKeywords`, `WithBases`, etc.): Takes an additional argument.
- `Flags`: Takes additional flags
- `Ex`: Takes additional arguments. Avoid this; use a more specific name.

XXX: Should we prefer adding these suffixes rather to using a different base name. (Should the "fixed" version of `PyDict_GetItem` be `PyDict_GetItemRef` or `PyDict_FetchItem`?) IMO, the suffix is uglier but clearer, and clarity wins that contest.

## Avoid

- `New`
- `Fast`
- `Ex`

**Comment by davidhewitt** [comment]
@encukou what do you think about the guideline for `PyInt` and `PyStr` extending to downstream projects? At the moment PyO3 is a little confused about what to name these types, we expose `PyInt` and `PyLong` as aliases for the same type, and similarly we have aliases `PyUnicode` and `PyString` (oops, not even `PyStr`).

Should PyO3 back up and reduce just back to `PyLong` and `PyUnicode`? Or do you think it's better for PyO3 to go for the clean break (because it can relatively easily deprecate things compared to CPython) and switch over to `PyInt` / `PyStr`?

**Comment by gvanrossum** [comment]
I’m not Petr but I think that’s really up to PyO3.

**Comment by davidhewitt** [comment]
@gvanrossum fair point and I've intended to start a thread on this on the PyO3 issue tracker for a while. For what it's worth I was leaning towards `PyInt` and `PyStr` because I think it's generally easier for PyO3 users to find APIs when they share names with pure Python code rather than the C API.

I'll mark as off-topic.

**Comment by encukou** [comment]
The question is off topic, but this discussion might be interesting for us as well.

Yup, it depends on the project.
One thing to watch out for is that Python reserves the `Py` namespace, so it's free to use `PyInt*` or `PyStr*` in the future. Don't extend the namespace with aliases to current functions. The naming guidelines we're coming up here are meant to change and adapt as time goes on.

But PyO3 seems have a separate `cffi` namespace, so it should be fine here.

**Comment by encukou** [comment]
Should we prefer adding suffixes rather to using a different base name? E.g. should the "fixed" version of ``PyDict_GetItem`` be ``PyDict_GetItemRef`` or ``PyDict_FetchItem``?

IMO, the suffix is uglier but clearer, and clarity wins that contest.
